### ISSUE TITLE
Offset for partialVisibility

### ISSUE THREAD

**Issue by eek** [joshwnj/react-visibility-sensor#53]
I think a good addition would be to add an offset to partialVisibility (required px from different sides in order for the element to be visible or not) it would be very useful for elements that are almost as big as the viewport.

Same as how the offset from jquery appear plugin works. I can try to implement and send a PR but after #41 is merged.

**Comment by roopemerikukka** [comment]
I'm not quite sure if this was what you were thinking but we have used this component and we needed an option to check if the sensor is lets say 100px below viewport bottom and we have set offsetTop for it to 110px it should be visible. I've made fork from the repo and my solution is here: https://github.com/roopemerikukka/react-visibility-sensor/tree/feature-offset (and here is the exact commit https://github.com/roopemerikukka/react-visibility-sensor/commit/7aa2c87216493e640dbe7e72b4e496385964b232)

Do you think this is something that would benefit people and I should make a PR from it?

It works like the partialVisibility but you can set offset to one direction at a time:
```jsx
<VisibilitySensor offset={{direction: 'top', value: 110}} />
```

I've only done manual testing so ofc. it still needs some tests to be written.

**Comment by sarukuku** [comment]
👍

**Comment by brandondurham** [comment]
@roopemerikukka — For sure.

**Comment by brandondurham** [comment]
@roopemerikukka — I tried your commit and may be doing something wrong.

I need my 84px tall element to register as `hidden` when the top reaches `20px` from the top of the viewport. I tried several options and nothing seemed to change the behavior.

```
<VisibilitySensor
	onChange={ this.onVisibility }
	offset={ {
		direction: 'top',
		value: 20
	} }
/>
```

**Comment by roopemerikukka** [comment]
@brandondurham  hmm this is working a bit differently. The offset value in my version tells that if the element is 20px above the viewport top it should still be visible. So I'm using it to tell that the element is "close to coming visible" so we can already assume it as visible.

I've just made couple of tests for my own case, but I can try to check if that could also work by giving negative values like `{value: -20}`. Now it doesn't as it converts the value always to positive with `Math.abs()` but may work only just by removing that.

**Comment by roopemerikukka** [comment]
Yep it works if you remove the abs. conversion and use negative value. I'll try to create the pull request by tomorrow.

**Comment by roopemerikukka** [comment]
@brandondurham should I though change it in a way that the positive values means inside the viewport and negatives outside? Do you think it makes more sense?

**Comment by sarukuku** [comment]
Any updates on this? I'd also like to get this feature in. 👍

**Comment by eek** [comment]
This is great @roopemerikukka ! And yeah, I think going with negative values as default makes more sense (if 20px offset then it's hidden if top is below <= 20px)

**Comment by sarukuku** [comment]
Could the #64 be now merged?

**Comment by joshwnj** [comment]
@sarukuku yes! :) Just merged and published a new version (3.7.0) along with a small refactor (#67).

Thanks to everyone on this thread

**Comment by sarukuku** [comment]
🎉

**Comment by jadus** [comment]
Great. But I'd like to set the offset for both bottom and top directions. How can I do this ?

**Comment by sarukuku** [comment]
@jadus did you try?
```javascript
<VisibilitySensor
	onChange={ this.onVisibility }
	offset={ {
		direction: 'bottom',
		value: 20
	} }
/>
```

If that doesn't work debug, fix and submit a PR. 😄

**Comment by falcon1kr** [comment]
Hello All, check out #65 . It's my take on the offset + partialVisibility and multiple offsets. Had to change the API to use { <direction>: <value> } instead of the current one.

**Comment by roopemerikukka** [comment]
Yep the current version doesn't work with multiple offsets. And as I noticed I accidentally used wrong comparison operators, so if you have element that has 100% width it doesn't count as visible because the `containmentRect.left = 0` and `rect.left = 0` but the offset compares them `containmentRect.left < rect.left` as it should be `containmentRect.left <= rect.left`. This can be avoided with a bit hacky way by defining the width of the element to be less than 100%:

```css
.el {
  width: 99.99%;
  margin: 0 auto;
}
```

I think that there is no need for me to fix this as there is better solution already in pull requests.
